### PR TITLE
refactor: refactor to use auth0 agnostic jwks package internally

### PR DESCRIPTION
**Pull request by ergusto**
### Introduction

Refactoring around the new auth0 agnostic jwt jwks package (https://github.com/nearform/fastify-jwt-jwks). 

**Changes**

* Adds fastify-jwt-jwks as a dependency
* Refactors index.js into a wrapper around the fastify-jwt-jwks plugin, constructing jwks url before passing directly to fastify-jwt-jwks
* Proxies the jwtJwks Fastify decorator exposed by fastify-jwt-jwks to auth0Verify
* Removes no longer required dependencies 
* Updates Readme to reflect refactor

### CODE REVIEW

**Comment by simoneb** [comment]
is this PR ready? (asking as I've been added as a reviewer, but it's not necessary until the PR is actually ready to be reviewed)

**Comment by ergusto** [comment]
> is this PR ready? (asking as I've been added as a reviewer, but it's not necessary until the PR is actually ready to be reviewed)

Ah that makes sense. Not ready yet, still some refactoring of tests etc to do, but had some meetings this afternoon. I'll add you again when it's ready.

File: test/test.js [hunk 1]
```diff
@@ -208,23 +208,23 @@ describe('Options parsing', function () {
   it('should enable RS256 when the domain is present', async function () {
     const server = await buildServer({ domain: 'localhost' })
 
-    expect(server.auth0Verify.verify.algorithms).toEqual(['RS256'])
+    expect(server.jwtJwks.verify.algorithms).toEqual(['RS256'])
```
[the remark this answers is not in the document]
Comment (ergusto) [reply test/test.js:211]: Good idea! I'll revert that commit and add a proxy from auth0Verify to jwtJwks.

[the remark this answers is not in the document]
Comment (simoneb) [reply test/test.js:211]: in the docs let's also point out that this implementation in now based off the other, I'm not sure if it's worth duplicating all the docs, I think we should keep them only in the new repo and only reference them from here, besides documenting what this plugin does specifically on top of the other

[the remark this answers is not in the document]
Comment (ergusto) [reply test/test.js]: @simoneb Updated the docs and added a proxy/reverted the previous commit.

File: package.json [hunk 2]
```diff
@@ -48,10 +48,9 @@
   "dependencies": {
     "@fastify/cookie": "^8.0.0",
     "@fastify/jwt": "^6.1.0",
+    "fastify-jwt-jwks": "github:nearform/fastify-jwt-jwks",
     "fastify-plugin": "^4.0.0",
-    "http-errors": "^2.0.0",
-    "node-cache": "^5.0.1",
-    "node-fetch": "^2.6.1"
+    "node-cache": "^5.0.1"
```
Comment (simoneb) [inline package.json:53]: do we still need this dependency?

File: index.js [hunk 3]
```diff
@@ -1,205 +1,40 @@
...
-      ttl > 0 ? new NodeCache({ stdTTL: ttl }) : { get: () => undefined, set: () => false, close: () => undefined }
-
-    // Create a cache or a fake cache
-    instance.decorateRequest('auth0VerifySecretsCache', {
-      getter: () => cache
-    })
+    const plugin = fastifyJwtJwks(instance, options, done)
 
-    instance.addHook('onClose', () => cache.close())
+    instance.decorate('auth0Verify', instance.jwtJwks)
```
Comment (simoneb) [inline index.js:35]: I'm slightly unsure about this implementation, the sequence of operations doesn't look entirely right. I mean plugin initialization is asynchronous, and we're using callbacks (we should probably switch to using promises). You're initializing the other plugin but without waiting you're decorating the instance with a decorator that, as far as we know, is not yet available. 

Even if this works, I would definitely consider making it more obvious that it works.

File: package.json (hunk 2 above)
Comment (ergusto) [reply package.json:53]: Ah no, I'll remove it.

File: index.js (hunk 3 above)
Comment (ergusto) [reply index.js]: @simoneb I've pushed a change that adds the decorator in an onReady hook. I believe this should resolve the issue and the tests are passing.

File: index.js [hunk 4]
```diff
@@ -1,205 +1,47 @@
...
+      throw new Error('Big scary error')
+    }
 
-    // Create a cache or a fake cache
-    instance.decorateRequest('auth0VerifySecretsCache', {
-      getter: () => cache
+    instance.addHook('onReady', function (done) {
+      instance.decorate('auth0Verify', instance.jwtJwks)
+      done()
     })
```
Comment (simoneb) [inline index.js:42]: I don't like much this implementation either. We can register our decorator when the callback of the other plugins is called, meaning that the other plugin has initialized and the decorator should be available.

Please watch out for binding issues in providing instance.jwtJwks directly in the decorator initialization.

I understand tests are passing but I wouldn't be too sure they're testing explicitly for these scenarios, as these are potentially novel issues related to the implementation change

File: index.js [hunk 5]
```diff
@@ -1,205 +1,43 @@
...
-      ttl > 0 ? new NodeCache({ stdTTL: ttl }) : { get: () => undefined, set: () => false, close: () => undefined }
+    return fastifyJwtJwks(instance, options, function (e) {
+      if (e) {
+        return done(e)
+      }
 
-    // Create a cache or a fake cache
-    instance.decorateRequest('auth0VerifySecretsCache', {
-      getter: () => cache
+      instance.decorate('auth0Verify', { ...instance.jwtJwks })
```
Comment (ergusto) [inline index.js:38]: @simoneb I wasn't sure what the best way to get around the binding issue would be, is this sufficient?

Comment (simoneb) [reply index.js:38]: I'd prefer to bind the function to the instance, unless fastify guarantees that already

Comment (ergusto) [reply index.js:38]: `jwtJwks` is just an options object with a few properties, no methods, so I'm not sure that applies.

Comment (simoneb) [reply index.js:38]: I see, then no need

**Review by simoneb — APPROVED** [review]
LGTM but before merging let's release the new package in the other repo first.

**Pull request merged**